dashboard/app/models/user.rb

```diff
@@ -963,11 +963,12 @@ def clever_student?
   # overrides Devise::Authenticatable#find_first_by_auth_conditions
   # see https://github.com/plataformatec/devise/blob/master/lib/devise/models/authenticatable.rb#L245
   def self.find_for_authentication(tainted_conditions)
+    max_credential_size = 255
     conditions = devise_parameter_filter.filter(tainted_conditions.dup)
     # we get either a login (username) or hashed_email
     login = conditions.delete(:login)
     if login.present?
-      return nil if login.utf8mb4?
+      return nil if login.size > max_credential_size || login.utf8mb4?
```

Comment:
It appears that the `username` attribute has validations (such as max length 20 characters and a regex that doesn't permit spaces).  Could we apply those same validations here?

Reply:
Could we skip the `username = :value` SQL comparison if login fails the regex and length validations?

Reply:
I'm hesitant to do that because we have user rows that don't pass those validations:
```
mysql> select max(length(username)) from users where id between 34000000 and 35000000;
+-----------------------+
| max(length(username)) |
+-----------------------+
|                    25 |
+-----------------------+
```

Reply:
Ah, good point.  In generally, I think it's worth spending web application server CPU to avoid async activity (long database queries).  Look's like that's not possible here without cleaning up existing usernames.